mender-update/deployments.cpp

```diff
@@ -0,0 +1,207 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <sstream>
+#include <string>
+
+#include <common/common.hpp>
+#include <common/error.hpp>
+#include <common/events.hpp>
+#include <common/expected.hpp>
+#include <common/http.hpp>
+#include <common/io.hpp>
+#include <common/json.hpp>
+#include <common/log.hpp>
+#include <common/optional.hpp>
+#include <common/path.hpp>
+#include <mender-update/context.hpp>
+
+#include <mender-update/deployments.hpp>
+
+namespace mender {
+namespace update {
+namespace deployments {
+
+using namespace std;
+
+namespace common = mender::common;
+namespace context = mender::update::context;
+namespace error = mender::common::error;
+namespace events = mender::common::events;
+namespace expected = mender::common::expected;
+namespace io = mender::common::io;
+namespace json = mender::common::json;
+namespace log = mender::common::log;
+namespace optional = mender::common::optional;
+namespace path = mender::common::path;
+
+const DeploymentsErrorCategoryClass DeploymentsErrorCategory;
+
+const char *DeploymentsErrorCategoryClass::name() const noexcept {
+	return "DeploymentsErrorCategory";
+}
+
+string DeploymentsErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	case InvalidDataError:
+		return "Invalid data error";
+	case BadResponseError:
+		return "Bad response error";
+	}
+	assert(false);
+	return "Unknown";
+}
+
+error::Error MakeError(DeploymentsErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, DeploymentsErrorCategory), msg);
+}
+
+const string v1_uri = "api/devices/v1/deployments/device/deployments/next";
+const string v2_uri = "api/devices/v2/deployments/device/deployments/next";
+
+error::Error CheckNewDeployments(
+	context::MenderContext &ctx,
+	const string &server_url,
+	http::Client &client,
+	events::EventLoop &loop,
+	APIResponseHandler api_handler) {
+	auto ex_dev_type = ctx.GetDeviceType();
+	if (!ex_dev_type) {
+		return ex_dev_type.error();
+	}
+	string device_type = ex_dev_type.value();
+
+	auto ex_provides = ctx.LoadProvides();
+	if (!ex_provides) {
+		return ex_provides.error();
+	}
+	auto provides = ex_provides.value();
+	if (provides.find("artifact_name") == provides.end()) {
+		return MakeError(InvalidDataError, "Missing artifact name data");
+	}
+	string artifact_name = provides["artifact_name"];
```

Comment:
Nit: const

Reply:
Yeah, this is always a thing in C/C++. But we don't do it much often in our code elsewhere...

Reply:
And have a rule that every variable containing data we don't want to modify should be declared `const`?

Reply:
If we mean local variables, I think I'd be against such a rule. Don't get me wrong, using const is good, please do whenever you see fit. But to force it onto all local variables as a rule is too much, IMHO.

At least before doing so we should consider what kind of mistakes we think this is preventing.

Reply:
I don't have enough experience with it to say, really.

But what I do know is this: I have never thought a local const variable has saved me from doing something stupid. It's already local, so I can easily see everywhere it's used.

OTOH I have often been annoyed at variables that are const, because I actually need to change them.

This is a democratic decision, so if both of your are in favor, we will do it. I just think a hard rule is a bit much.

Reply:
Yes, I agree. A hard rule is probably a bit much, but I do think we should keep it in mind.

Seems the style guide is also not giving us much here:
https://google.github.io/styleguide/cppguide.html#Use_of_const

```Using const on local variables is neither encouraged nor discouraged.```

Reply:
> That other language like Rust for instance have const as the default doesn't change ur mind?

Not everything Rust does is good and the right way :wink:

Reply:
> This is a democratic decision, so if both of your are in favor, we will do it. I just think a hard rule is a bit much.

I'm definitely not in favor of a hard rule. Generally, I still believe programming is a craft and the programmer is responsible for their creation and they create it in a way that works for them. `const` local variables sometimes help with readability and to emphasize something, but if used everywhere, it's just noise, I think.

Reply:
Ok, it sounds like we pretty much agree. Local variable const is good to keep in mind; use it where it makes sense, but no hard rule.